### PR TITLE
Improve sd calculation

### PR DESCRIPTION
**Pull request by ben-heil**
Improved logic to handle a lot of edge cases, added empirical p_value calculation in the case of a zero standard deviation, and added some tests.

Addresses issues mentioned [here](https://github.com/greenelab/hetmech/issues/153)

### CODE REVIEW

File: hetmatpy/pipeline.py [hunk 1]
```diff
@@ -26,12 +40,38 @@ def add_gamma_hurdle_to_dgp_df(dgp_df):
...
+    Calculate the p_value for a given metapath if possible
+    """
+    if row['sum'] == 0:
+        return None
+    elif row['path_count'] == 0:
+        return 1.0
+    # If the standard deviation is zero, calculate the p_value empiricly
+    elif row['sd_nz'] == 0:
+        if row['dwpc'] <= dgp_df['mean_nz']:
+            return dgp_df['nnz'] / dgp_df['n_dwpcs']
```
Comment (ben-heil) [inline hetmatpy/pipeline.py]: I'm pretty sure these are the two correct fields for calculating the p_value, but I'd appreciate someone double checking it

**Comment by ben-heil** [comment]
I added some changes to the PR to address some comments Daniel had when we met last Thursday.  

One thing that I found [at the end of this notebook](https://github.com/ben-heil/hetmech/blob/307f5caa7326bcb71e058ca9d3d6e6dde81a66a1/small_standard_deviations.ipynb) was that the number of nonzero permutations is not correlated with the spurious standard deviations. As a result, we should be able to continue using a constant error tolerance in calculating the standard deviation. 


image

File: hetmatpy/pipeline.py [hunk 2]
```diff
@@ -8,6 +8,22 @@
...
+
+def calculate_sd(sum_of_squares, unsquared_sum, number_nonzero):
+    """
+    Calculate the standard deviation and validate the incoming data
+    """
+    if number_nonzero == 1:
+        return 0
+    # If all the values in the row are the same we'll manually return zero,
+    # because not doing so can lead to some issues with float imprecision
+    elif abs(sum_of_squares - unsquared_sum ** 2 / number_nonzero) < FLOAT_ERROR_TOLERANCE:
```
Comment (dhimmel) [inline hetmatpy/pipeline.py]: Should you assign this value to `numerator` or something to prevent the duplicate computation below? These functions get called a LOT, so efficiency is a consideration

File: hetmatpy/pipeline.py [hunk 3]
```diff
@@ -8,6 +8,22 @@
...
 
+FLOAT_ERROR_TOLERANCE = 1e-5
+
+
+def calculate_sd(sum_of_squares, unsquared_sum, number_nonzero):
+    """
+    Calculate the standard deviation and validate the incoming data
+    """
+    if number_nonzero == 1:
+        return 0
```
Comment (dhimmel) [inline hetmatpy/pipeline.py]: My understanding is that this should be None / NaN due to divide by zero https://stackoverflow.com/a/8023909/4651668

Why break it out as a special case?

File: hetmatpy/pipeline.py [hunk 4]
```diff
@@ -26,12 +42,51 @@ def add_gamma_hurdle_to_dgp_df(dgp_df):
...
+
+
+def calculate_empirical_p_value(row, dgp_df):
+    """
+    Calculate p_value in cases where the gamma hurdle model won't work
+    """
+    if row['nnz'] == 0:
+        return 0
+    elif row['path_count'] == 0:
+        return 1.0
```
Comment (dhimmel) [inline hetmatpy/pipeline.py]: This should have higher priority than `if row['nnz'] == 0`?

File: hetmatpy/pipeline.py [hunk 5]
```diff
@@ -26,12 +42,51 @@ def add_gamma_hurdle_to_dgp_df(dgp_df):
         )
     # Compute gamma-hurdle parameters
     dgp_df['mean_nz'] = dgp_df['sum'] / dgp_df['nnz']
```
Comment (dhimmel) [inline hetmatpy/pipeline.py:50]: Did this have any precision issues where `mean_nz` was nonzero when it should be zero? I guess that would depend on whether there are any issues with `dgp_df['sum']` being nonzero when it should be zero?

File: hetmatpy/pipeline.py (hunk 1 above)
Comment (dhimmel) [reply hetmatpy/pipeline.py]: Yes correct.

File: hetmatpy/pipeline.py [hunk 6]
```diff
@@ -26,12 +42,51 @@ def add_gamma_hurdle_to_dgp_df(dgp_df):
...
+def calculate_gamma_hurdle_p_value(row):
+    """
+    Use the gamma hurdle model to calculate the p_value for a metapath
+    """
+    return row['nnz'] / row['n'] * (scipy.special.gammaincc(row['alpha'], row['beta'] * row['dwpc']))
+
+
+def calculate_empirical_p_value(row, dgp_df):
+    """
+    Calculate p_value in cases where the gamma hurdle model won't work
```
Comment (dhimmel) [inline hetmatpy/pipeline.py]: I would like comments in this function since the logic flow can get confusing.

File: hetmatpy/pipeline.py (hunk 3 above)
Comment (ben-heil) [reply hetmatpy/pipeline.py]: I pulled it out as a special case to prevent division by zero in calculating the standard deviation.

You're right though, it should be returning None instead of 0. I'll update this, and modify the code in add_gamma_hurdle to look for standard deviations of None in addition to 0.

File: hetmatpy/pipeline.py (hunk 5 above)
Comment (ben-heil) [reply hetmatpy/pipeline.py:50]: It doesn't appear to be the case, the smallest mean_nz in the database is .15, and seems pretty evenly distributed up from there.

File: hetmatpy/pipeline.py [hunk 7]
```diff
@@ -26,12 +48,56 @@ def add_gamma_hurdle_to_dgp_df(dgp_df):
...
+    Use the gamma hurdle model to calculate the p_value for a metapath
+    """
+    return row['nnz'] / row['n'] * (scipy.special.gammaincc(row['alpha'], row['beta'] * row['dwpc']))
+
+
+def calculate_empirical_p_value(row):
+    """
+    Calculate p_value in cases where the gamma hurdle model won't work
+    """
+    if row['path_count'] == 0:
```
Comment (ben-heil) [inline hetmatpy/pipeline.py]:
```
if row['path_count'] == 0 if 'path_count' in row else row['dwpc'] == 0.0:
```

NVM

File: hetmatpy/tests/test_pipeline.py [hunk 8]
```diff
@@ -1,6 +1,12 @@
 from hetmatpy.pipeline import (
     grouper,
+    calculate_sd,
+    calculate_p_value,
+    add_gamma_hurdle_to_dgp_df
 )
+import numpy
+import pandas
+import pytest
```
Comment (ben-heil) [inline hetmatpy/tests/test_pipeline.py]: Move imports above

File: hetmatpy/tests/test_pipeline.py [hunk 9]
```diff
@@ -46,3 +52,66 @@ def test_grouper_length_1():
...
+    # Only one nonzero observation, so return zero
+    (5, 5, 1, None),
+    # Test that the standard deviation of 5, 4, and 3 is 1
+    (50, 12, 3, 1)])
+def test_calculate_sd(sum_of_squares, unsquared_sum, number_nonzero, expected_output):
+    assert calculate_sd(sum_of_squares, unsquared_sum, number_nonzero) == expected_output
+
+
+@pytest.mark.parametrize('row, expected_output', [
+    # Zero path count
```
Comment (ben-heil) [inline hetmatpy/tests/test_pipeline.py:74]: floatify relevent fields

**Review by dhimmel — APPROVED** [review]
